nixos/modules/misc/wordlist.nix

```
@@ -0,0 +1,65 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  /* Processes the list of paths Returns either the sole path passed in if it's
+     a singleton list.  Otherwise creates a wordlist constructed from the
+     various paths. */
+  processList = name: paths:
+    if builtins.length paths == 1
+      then head paths
+      else pkgs.concatText name paths;
+
+in
+{
+  options = {
+    environment.wordlist = {
+      enable = mkEnableOption "environment variables for lists of words";
+
+      lists = mkOption {
+        type = types.attrsOf (types.nonEmptyListOf types.path);
+
+        default = {
+          WORDLIST = [ "${pkgs.scowl}/share/dict/words.txt" ];
+        };
+
+        defaultText = literalExpression ''
+          {
+            WORDLIST = [ "''${pkgs.scowl}/share/dict/words.txt" ];
+          }
+        '';
+
+        description = ''
+          A set with the key names being the environment variable you'd like to
+          set and the values being a list of paths to text documents containing
+          lists of words. Path lists containing a single path will just use the
+          path to that value directly in the environment variable, whereas
+          multiple paths will create a concatenation of the files and store that
+          in the nix store.
+
+          Please note this behavior if you are referencing a file outside of the
+          nix store.  Such files may not be visible to your users.  To
+          workaround this, simply create a derivation containing that file and
+          reference that instead. You may find `pkgs.writeText` useful for this
+          task.
+        '';
+
+        example = literalExpression ''
+          {
+            WORDLIST = [ "''${pkgs.scowl}/share/dict/words.txt" ];
+            AUGMENTED_WORDLIST = [
+              "''${pkgs.scowl}/share/dict/words.txt"
+              "''${pkgs.scowl}/share/dict/words.variants.txt"
+            ];
+          }
+        '';
+      };
+    };
+  };
+
+  config = mkIf config.environment.wordlist.enable {
+    environment.variables =
+      lib.mapAttrs
+        (name: value: builtins.toString (processList "wordlist-${name}" value))
```

we shouldn't write reviews so close to bedtime :sweat: `toString` is of course not sufficient because paths don't have context, so `toString ./list == "/path/to/the/list"` and nothing is copied into the store. we actually need
```suggestion
        (name: value: "${processList "wordlist-${name}" value}")
```
to copy contents of path literals into the store so that all users can access them. in fact we may even want to use `concatText` in all cases, even lists of length 1, because that doesn't require us keeping the full scowl package around: the word list gets copied out during system build, and scowl can be GCd afterwards to get rid of parts of it that aren't actually in use.

---

it's okay to have this behaviour if the user put a stringy path into the config (eg `[ "/etc/words" ]` or so). using path literals should always copy the referenced file into the store though, since that's what's generally expected to happen when using path literals. :)